lib/rubygems/installer.rb

```diff
@@ -676,7 +676,14 @@ def process_options # :nodoc:
     @build_args = options[:build_args]
 
     @gem_home = @install_dir
-    @gem_home ||= options[:user_install] ? Gem.user_dir : Gem.dir
+    @gem_home ||= if options[:user_install]
+      Gem.user_dir
+    elsif !ENV.key?("GEM_HOME") && (File.exist?(Gem.dir) && !File.writable?(Gem.dir))
+      say "Defaulting to user installation because default installation directory (#{Gem.dir}) is not writable."
+      Gem.user_dir
+    else
+      Gem.dir
+    end
```

Comment:
Adding this change will fix `--build-root` as well.

```suggestion
    @gem_home = @install_dir || Gem.dir

    # `--build-root` overrides `--user-install` and auto-user-install
    if @build_root.nil?
      if options[:user_install]
        @gem_home = Gem.user_dir
      elsif !ENV.key?("GEM_HOME") && (File.exist?(Gem.dir) && !File.writable?(Gem.dir))
        say "Defaulting to user installation because default installation directory (#{Gem.dir}) is not writable."
        @gem_home = Gem.user_dir
      end
    end
```

Reply:
Yep, that makes sense to me! I'll work on complementing your fix with a test tomorrow!

Reply:
Made a slight edit to my suggestion (added `!options[:install_dir] && `) because I had accidentally broken `--install-dir` taking priority over `--user-install`.

Reply:
Marking as resolved, in favor of the patch (with a test) I mentioned at <https://github.com/rubygems/rubygems/pull/7212#issuecomment-1835268767>.